Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp

```diff
@@ -0,0 +1,292 @@
+	if (MaxDistanceBetweenPlayers >= PlayerCheckoutRadius)
+	{
+		UE_LOG(LogBenchmarkGym, Error, TEXT("Maximum distance between players is larger than checkout radius. Cannot ensure players will always see each other."));
+	}
+
+	for (int i = 0; i < SpawnPointsNum; i++)
+	{
+		const int Row = i % NumRows;
+		const int Col = i / NumRows;
+
+		const int X = CenterX + MinRelativeX + Col * DistBetweenSpawnPoints;
+		const int Y = CenterY + MinRelativeY + Row * DistBetweenSpawnPoints;
+
+		FActorSpawnParameters SpawnInfo{};
+		SpawnInfo.Owner = this;
+		SpawnInfo.Instigator = NULL;
+		SpawnInfo.bDeferConstruction = false;
+		SpawnInfo.SpawnCollisionHandlingOverride = ESpawnActorCollisionHandlingMethod::AlwaysSpawn;
+
+		const FVector SpawnLocation = FVector(X, Y, Z);
+		UE_LOG(LogBenchmarkGym, Log, TEXT("Creating a new PlayerStart at location %s."), *SpawnLocation.ToString());
+		SpawnPoints.Add(World->SpawnActor<APlayerStart>(APlayerStart::StaticClass(), SpawnLocation, FRotator::ZeroRotator, SpawnInfo));
+	}
+}
+
+void ABenchmarkGymGameMode::SpawnNPCs(int NumNPCs)
+{
+	for (int32 i = 0; i < NumNPCs; i++)
+	{
+		int32 Cluster = i % NumPlayerClusters;
+		int32 SpawnPointIndex = Cluster * PlayerDensity;
+		AActor* SpawnPoint = SpawnPoints[SpawnPointIndex];
+		SpawnNPC(SpawnPoint->GetActorLocation());
+	}
+}
+
+void ABenchmarkGymGameMode::SpawnNPC(const FVector& SpawnLocation)
+{
+	UWorld* const World = GetWorld();
+	if (World == nullptr)
+	{
+		UE_LOG(LogBenchmarkGym, Error, TEXT("Error spawning NPC, World is null"));
+		return;
+	}
+
+	if (NPCPawnClass == nullptr)
+	{
+		UE_LOG(LogBenchmarkGym, Error, TEXT("Error spawning NPC, NPCPawnClass is not set."));
+		return;
+	}
+
+	FVector RandomOffset = RNG.GetUnitVector()*600.0f;
+	if (RandomOffset.Z < 0.0f) 
+		RandomOffset.Z = -RandomOffset.Z;
+	FVector RandomSpawn = SpawnLocation + RandomOffset;
+	UE_LOG(LogBenchmarkGym, Log, TEXT("Spawning NPC at %s"), *SpawnLocation.ToString());
+	FActorSpawnParameters SpawnInfo{};
+	//SpawnInfo.Owner = this;
+	//SpawnInfo.Instigator = NULL;
+	SpawnInfo.SpawnCollisionHandlingOverride = ESpawnActorCollisionHandlingMethod::AlwaysSpawn;
+	World->SpawnActor<APawn>(NPCPawnClass->GetDefaultObject()->GetClass(), RandomSpawn, FRotator::ZeroRotator, SpawnInfo);
+}
+
+void ABenchmarkGymGameMode::StartPlay()
+{
+	UE_LOG(LogBenchmarkGym, Log, TEXT("Enabling custom density spawning."));
+	ParsePassedValues();
+	ClearExistingSpawnPoints();
+
+	SpawnPoints.Reset();
+	GenerateSpawnPointClusters(NumPlayerClusters);
+
+	if (SpawnPoints.Num() != TotalPlayers) 
+	{
+		UE_LOG(LogBenchmarkGym, Error, TEXT("Error creating spawnpoints, number of created spawn points (%d) does not equal total players (%d)"), SpawnPoints.Num(), TotalPlayers);
+	}
+
+	SpawnNPCs(TotalNPCs);
+	return Super::StartPlay();
+}
+AActor* ABenchmarkGymGameMode::FindPlayerStart_Implementation(AController* Player, const FString& IncomingName)
```

Comment:
Maybe AController* Player can be const? FindPlayerStart_Implementation sounds is if it might take a const pointer.